File: binary-tree-level-order-traversal-ii/binary-tree-level-order-traversal-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> res;
        DFS(root, 0, res);
        reverse(res.begin(), res.end());
        return res;
    }
    void DFS(TreeNode* node, int depth, vector<vector<int>>& res){
        if (!node) return;
        if (res.size() <= depth) res.push_back({});
        res[depth].push_back(node->val);
        DFS(node->left, depth + 1, res);
        DFS(node->right, depth + 1, res);
    }
};

File: binary-tree-level-order-traversal-ii/binary-tree-level-order-traversal-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrderBottom(self, root: Optional[TreeNode]) -> List[List[int]]:
        # BFS approach and return reverse list
        if not root: return []
        res = []
        curr = [root]
        
        while curr:
            next_ = []
            res.append([])
            for node in curr:
                res[-1].append(node.val)
                if node.left: next_.append(node.left)
                if node.right: next_.append(node.right)
            curr = next_
        return res[::-1]